lib/chef/telemetry/decision.rb

```diff
@@ -1,12 +1,91 @@
+    class Decision
+      extend Forwardable
+
+      attr_reader :config, :enabled, :arg_decision, :env_decision, :file_decision, :prompt_decision, :logger
+
+      def initialize(opts = {})
+        @config = opts
+        @logger = opts[:logger] || Logger.new(opts.key?(:output) ? opts[:output] : STDERR)
+        @config[:output] ||= STDOUT
+        config[:logger] = logger
+
+        # This is the whole point - whether telemetry shoud be enabled in this
+        # runtime invocation. Start by assuming no.
+        @enabled = false
+
+        # The various things that have a say in whether telemetry should be enabled.
+        @arg_decision = Decision::Argument.new(ARGV)
+        @env_decision = Decision::Environment.new(ENV)
+        @file_decision = Decision::File.new(config)
+        @prompt_decision = Decision::Prompt.new(config)
+      end
+
+      #
+      # Methods for obtaining consent from the user.
+      #
+      def check_and_persist(dir = ".")
+        file_decision.local_dir = dir
+
+        # If a decision is made by CLI arg, set runtime decision and do not persist
+        logger.debug "Telemetry decision examining CLI arg checks"
+        return @enabled = true if @arg_decision.enable?
+        return @enabled = false if @arg_decision.disable?
+
+        # If a non-persisting decision is made by env, only set runtime decision
+        logger.debug "Telemetry decision examining ephemeral ENV checks"
+        return @enabled = true if @env_decision.opt_in_no_persist?
+        return @enabled = false if @env_decision.opt_out_no_persist?
+
+        # Check to see if a persistent decision is made by env but not yet persisted
+        #  then persist it and set runtime decision
+        logger.debug "Telemetry decision examining persistent ENV checks"
+        if env_decision.opt_in?
+          file_decision.persist(true, dir) if !persisted? || file_decision.opt_out?
+          return @enabled = true
+        elsif env_decision.opt_out?
+          file_decision.persist(false, dir) if !persisted? || file_decision.opt_in?
+          return @enabled = false
+        end
+
+        # If a decision has been made by file, read from disk and set runtime decision
+        logger.debug "Telemetry decision examining file checks"
+        if persisted?
+          return @enabled = true if file_decision.opt_in?
+          return @enabled = false if file_decision.opt_out?
+        end
+
+        # Lowest priority is to interactively prompt if we have a TTY
+        if config[:output].isatty
+          logger.debug "Telemetry decision - detected TTY, prompting..."
+          return @enabled = prompt_decision.prompt(dir, file_decision)
+        end
+
+        # Otherwise no decision has been made, default to runtime opt-out
+        logger.debug "Telemetry decision - no decision, defaulting to opt-out"
+        @enabled = false
+      end
+
+      def self.check_and_persist(opts = {})
+        new(opts).check_and_persist
```

Comment:
Should this accept a dir to pass into `#check_and_persist?` for local dir, and/or allow that as an option that we handle?

Reply:
Thanks, that's a good catch - those args mismatch.